### ISSUE TITLE
reduce_sum in tflite outputs is incorrect

### ISSUE THREAD

**Issue by reuvenperetz** [tensorflow/tensorflow#50992]
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 18.04
- TensorFlow installation (pip package or built from source): pip
- TensorFlow library (version, if pip package or github SHA, if built from source): 2.5.0

### 2. Code

Provide code to help us reproduce your issues using one of the following options:

``` python
import tensorflow as tf
import numpy as np


keras = tf.keras
layers = keras.layers


def infer(tflite_model, img):
    interpreter = tf.lite.Interpreter(model_content=tflite_model)
    interpreter.allocate_tensors()
    input_details = interpreter.get_input_details()[0]
    interpreter.set_tensor(input_details["index"], img)
    interpreter.invoke()
    output_details = interpreter.get_output_details()[0]
    output = interpreter.get_tensor(output_details["index"])
    return output

def convert_to_tflite(model):
    converter = tf.lite.TFLiteConverter.from_keras_model(model)
    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    converter.experimental_new_converter = True
    tflite_model = converter.convert()
    return tflite_model


# Image to test
np.random.seed(1)
input_shape = (12, 14, 18)
img = np.random.random((1,)+input_shape).astype(np.float32) * 7.0 - 3.5

# Create a model
i = layers.Input(shape=input_shape)
x = tf.quantization.fake_quant_with_min_max_args(i, min=0.0, max=3.984375)
x = tf.reduce_sum(x, axis=None)
x = tf.quantization.fake_quant_with_min_max_args(x, min=-4.0, max=3.96875)
model = keras.models.Model(inputs=i, outputs=x)


tf_output = model.predict(img)
tflite_model = convert_to_tflite(model)
tflite_output = infer(tflite_model, img)

print(f'Tensorflow output: {tf_output}, TFLite output: {tflite_output}, ')

```

### 3. Failure after conversion
If the conversion is successful, but the generated model is wrong, then state what is wrong:
I get a huge difference between the TF model and the TFLite model:
Tensorflow output: 3.96875, TFLite output: -2.78125, 

The Tensorflow model gives completely different results than the TFLite model. Furthermore, before the reduce_sum, I quantize the tensor using only positive values, so it does not make sense that the output of the model is negative (which is what I get).

**Comment by rino20** [comment]
@daverim Could you take a look?

**Comment by daverim** [comment]
Hi, the reason why this happens is because tflite assumes that min/max ranges have been calculated correctly. If you change your example to `x = tf.quantization.fake_quant_with_min_max_args(x, min=0.0, max=781756.5)`, you'll see the two answers converge. In TF case, the reason why it works (more) correctly is that quant/dequant is entirely in float and uses min, max and rounding operations, whereas TFlite generally will avoid operations where possible to optimize latency. In TFLite's reduce sum kernel, there is a cast from int -> float for latency reasons (assuming that the scales are correct) vs rounding. The casting of wrong values can lead to negatives.

**Comment by reuvenperetz** [comment]
Hi,
Thanks for the answer!
But I'm not sure I fully got it... Could you please refer me to where this casting/assumption on correct min/max ranges is in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/reduce.cc? 

Thanks again :)

**Comment by reuvenperetz** [comment]
Hi @daverim,
When I remove the second fake-quant layer from the model, I get similar results.
Does the assumption on correct min/max ranges hold for all fake-quant layers in the model or just the one that quantizes the input?
And if so (it holds for all fake-quant layers), what happens if I trained my model poorly so the ranges are not correct? Does that mean that in this case, the layer's output doesn't follow the equations in https://www.tensorflow.org/api_docs/python/tf/quantization/fake_quant_with_min_max_args?

Also, does that mean that fake-quant layers should not be added to the model manually?

Thanks!

**Comment by reuvenperetz** [comment]
Hi,
I tried to use the same example but with different ranges, and I also get a huge difference between the TF output and TFLite output. The code is below, and the output is: "Tensorflow output: 4.0, TFLite output: 1.6627452373504639"
This time, the min/max of the first fake quant is matching the input distribution.
The second fake quant layer should only clip the single positive value to the maximum of the second fake quant, but I get 1.66 instead.

```python
import tensorflow as tf
import numpy as np


keras = tf.keras
layers = keras.layers


def infer(tflite_model, img):
    interpreter = tf.lite.Interpreter(model_content=tflite_model)
    interpreter.allocate_tensors()
    input_details = interpreter.get_input_details()[0]
    interpreter.set_tensor(input_details["index"], img)
    interpreter.invoke()
    output_details = interpreter.get_output_details()[0]
    output = interpreter.get_tensor(output_details["index"])
    return output

def convert_to_tflite(model):
    converter = tf.lite.TFLiteConverter.from_keras_model(model)
    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    converter.experimental_new_converter = True
    tflite_model = converter.convert()
    return tflite_model


# Image to test
np.random.seed(1)
input_shape = (12, 14, 18)
img = np.random.random((1,)+input_shape).astype(np.float32) * 7.0 - 3.5
img = np.clip(img, 0, 3)

# Create a model
i = layers.Input(shape=input_shape)
x = tf.quantization.fake_quant_with_min_max_args(i, min=0, max=4)
x = tf.reduce_sum(x, axis=None)
x = tf.quantization.fake_quant_with_min_max_args(x, min=0, max=4)
model = keras.models.Model(inputs=i, outputs=x)

print(np.max(img), np.min(img))
tf_output = model.predict(img)
tflite_model = convert_to_tflite(model)
tflite_output = infer(tflite_model, img)

print(f'Tensorflow output: {tf_output}, TFLite output: {tflite_output} ')
```

**Comment by mohantym** [comment]
I could replicate this issue in 2.8 version. Attached [gist](https://colab.sandbox.google.com/gist/mohantym/c8ca807113fd546e2c7750fd543afeed/git_50992.ipynb#scrollTo=WtW1EnfJwqra) for reference. Thank you!

**Comment by pkgoogle** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The TFLite team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the TFLite space.

**Comment by oawxkw** [comment]
I also observed the following API aliases or similar APIs can cause the same issue in older versions of tensorflow.
Users should be cautious when using them on both CPU and GPU from tensorflow 2.5.0 (v2.5.0-rc3-213-ga4dfb8d1a71) to 2.12.0 (v2.12.0-rc1-12-g0db597d0d75).

- `tf.reduce_sum`, `tf.compat.v1.reduce_sum`, `tf.compat.v1.math.reduce_sum`, `tf.math.reduce_sum`

It seems to be fixed in tensorflow 2.13.0 (v2.13.0-rc2-7-g1cb1a030a62) and later versions.
